x/attribute/keeper/keeper.go

```diff
@@ -382,39 +382,50 @@ func (k Keeper) DeleteAttribute(ctx sdk.Context, addr string, name string, value
 	}
 
 	store := ctx.KVStore(k.storeKey)
-	it := sdk.KVStorePrefixIterator(store, types.AddrStrAttributesNameKeyPrefix(addr, name))
-	defer it.Close()
-	var count int
-	for ; it.Valid(); it.Next() {
+	iter := sdk.KVStorePrefixIterator(store, types.AddrStrAttributesNameKeyPrefix(addr, name))
+	defer func() {
+		if iter != nil {
+			iter.Close()
+		}
+	}()
+
+	attrToDelete := []types.Attribute{} // do delete logic outside of iterator
+	for ; iter.Valid(); iter.Next() {
 		attr := types.Attribute{}
-		if err := k.cdc.Unmarshal(it.Value(), &attr); err != nil {
+		if err := k.cdc.Unmarshal(iter.Value(), &attr); err != nil {
 			return err
 		}
 
 		if attr.Name == name && (!deleteDistinct || bytes.Equal(*value, attr.Value)) {
-			count++
-			store.Delete(it.Key())
-			k.DecAttrNameAddressLookup(ctx, attr.Name, attr.GetAddressBytes())
-			k.deleteAttributeExpireLookup(store, attr)
+			attrToDelete = append(attrToDelete, attr)
+		}
+	}
+	iter.Close()
+	iter = nil
 
-			if !deleteDistinct {
-				deleteEvent := types.NewEventAttributeDelete(name, addr, owner.String())
-				if err := ctx.EventManager().EmitTypedEvent(deleteEvent); err != nil {
-					return err
-				}
-			} else {
-				deleteEvent := types.NewEventDistinctAttributeDelete(name, string(*value), addr, owner.String())
-				if err := ctx.EventManager().EmitTypedEvent(deleteEvent); err != nil {
-					return err
-				}
+	for _, attr := range attrToDelete {
+		addrBz := attr.GetAddressBytes()
+		store.Delete(types.AddrAttributeKey(addrBz, attr))
+		k.DecAttrNameAddressLookup(ctx, attr.Name, addrBz)
+		k.deleteAttributeExpireLookup(store, attr)
+		if !deleteDistinct {
+			deleteEvent := types.NewEventAttributeDelete(name, addr, owner.String())
+			if err := ctx.EventManager().EmitTypedEvent(deleteEvent); err != nil {
+				return err
+			}
+		} else {
+			deleteEvent := types.NewEventDistinctAttributeDelete(name, string(*value), addr, owner.String())
+			if err := ctx.EventManager().EmitTypedEvent(deleteEvent); err != nil {
+				return err
 			}
 		}
 	}
+
 	errm := "no keys deleted"
-	if count == 0 && deleteDistinct {
+	if len(attrToDelete) == 0 && deleteDistinct {
 		ctx.Logger().Error(errm, "name", name, "value")
 		return fmt.Errorf("%s with name %s value %s", errm, name, string(*value))
```

Comment:
We probably want `%q` for the `string(*value)`.

```suggestion
		return fmt.Errorf("%s with name %s value %q", errm, name, string(*value))
```